pysph/base/octree.pyx

```diff
@@ -450,127 +465,180 @@ cdef class Octree:
-
-                # Necessary to ensure i,j,k are treated as private variables
-                i = j = k = -1
-                for p from start<=p<(start+num_p):
-                    q = p_indices[p]
-                    find_cell_id_raw(
-                        src_x_ptr[q] - xmin[0],
-                        src_y_ptr[q] - xmin[1],
-                        src_z_ptr[q] - xmin[2],
-                        length/2,
-                        &i, &j, &k
-                        )
-                    
-                    oct_id = k+2*j+4*i
-                    (new_indices[oct_id]).push_back(q)
-                    hmax_children[oct_id] = fmax(hmax_children[oct_id], src_h_ptr[q])
-                eps = 2*self._get_eps(length, xmin)
-                length_padded = (length/2)*(1 + 2*eps)
-
-                c = start
-                for i from 0<=i<2:
-                    for j from 0<=j<2:
-                        for k from 0<=k<2:
-                            oct_id = k+2*j+4*i
- 
-                            if new_indices[oct_id].empty():
-                                continue
-
-                            xmin_new[0] = xmin[0] + (i - eps)*length/2
-                            xmin_new[1] = xmin[1] + (j - eps)*length/2
-                            xmin_new[2] = xmin[2] + (k - eps)*length/2
-
-                            node.children[oct_id] = self._new_node(xmin_new, length_padded,
-                                    hmax=hmax_children[oct_id], level=level+1, parent=node)
-                            node.children[oct_id].start_index = c
-                            node.children[oct_id].num_particles = new_indices[oct_id].size()
 
-                            # Change the position of the indices in p_indices 
-                            # according to which child they belong to
-                            for l from 0<=l<new_indices[oct_id].size():
-                                p_indices[c+l] = new_indices[oct_id][l]   
-                            c = c + new_indices[oct_id].size()
-
-
-        cdef vector[cOctreeNode *]* next_level_nodes = new vector[node_ptr]()
+            tid = threadid()
 
-        # All non-NULL nodes & non-leaf nodes are added to next_level_nodes, to process in next level
-        for i from 0<=i<level_nodes.size():
-            for j from 0<=j<8:
-                if (deref(level_nodes)[i].children[j] != NULL):
-                    node = deref(level_nodes)[i].children[j] 
-                    eps = 2*self._get_eps(node.length, node.xmin)
-                    if (node.num_particles < self.leaf_max_particles) or (eps > EPS_MAX):
-                        node.is_leaf = True
-                        continue
-                    next_level_nodes.push_back(node)
-
-        del level_nodes
+            while(count[num_threads] > 0):
+                for i from 0<=i<8:
+                    new_indices[i] = vector[u_int]()
+                children = 0
+
+                for n in xrange(old_nodes[tid].size()):
```

Comment:
Why are you using this `n in xrange`, instead why not use the `for i from ...` form or even just for i in range(...). I am not sure if this will be optimal?